### PR TITLE
feat(core): Add support for user defined types

### PR DESCRIPTION
**Pull request by RPate97**
## Purpose
Adds support for user-defined types

## Implementation Details
- Since the storage layout doesn't include information on the underlying type of user-defined types by default, I had to fetch the underlying type from the compiler output and inject it into the storage layout. 
- Note that while this feature adds support for user-defined types, we are limited to treating them like their underlying types.

### CODE REVIEW

**Comment by sam-goldman** [comment]
tl;dr I think there are a couple issues with this PR so I opened a new PR to be merged into this one.

The issues:

1. I don't think it's a good idea to modify the `StorageLayout` in-place, which is what we currently do in `addUserDefinedTypesToStorageLayout`. The storage layout object is attached to the build info, which could be used later in a totally different part of the codebase (e.g. when displaying the diff).

2. [Here](https://github.com/chugsplash/chugsplash/blob/bca4ecdf3a49e5a416d2966e1a5eeab6ec9a36ac/packages/core/src/actions/bundle.ts#L359), you only iterate over the nodes for the current contract, and not any of its parents. To illustrate why this is an issue, the following example will result in an error:

a. Go to `Storage.sol` in the plugins package

b. Move the type definitions from the `Storage` contract to a new parent contract, `ParentStorage`:
```
// New contract
contract ParentStorage {
    // User-defined types are in this parent contract
    type UserDefinedType is uint256;
    type UserDefinedBytes32 is bytes32;
    type UserDefinedInt is int;
    type UserDefinedInt8 is int8;
    type UserDefinedUint8 is uint8;
    type UserDefinedBool is bool;
}

// Storage now inherits from ParentStorage
contract Storage is ParentStorage {
    // Everything stays the same below
    enum TestEnum { A, B, C }
    ...
```
c. Run `yarn test:hardhat`

To fix this bug, we actually need to iterate over every source in the compiler output, since the user-defined type node can be in any one of them. Unfortunately, this gets error prone because it's tricky to guarantee you've found the right node, and it's slow to iterate over every node in the compiler output many times. Fortunately, there's a handy library made by OpenZeppelin called [solidity-ast](https://github.com/OpenZeppelin/solidity-ast) that's built exactly for this.

To solve these issues, I opened a new PR to be merged into this one. It uses the `solidity-ast` library and gets the underlying storage type on the fly instead of modifying the existing storage layout.

File: docs/variables.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ This is a reference that explains how to assign values to every variable type in
 - [Arrays](#arrays)
 - [Structs](#structs)
 - [Mappings](#mappings)
+- [User Defined Types](#user-defined-types)
```
Comment (sam-goldman) [inline docs/variables.md:21]: nit: User Defined Types -> User-Defined Value Types

File: docs/variables.md [hunk 2]
```diff
@@ -313,3 +314,17 @@ myMultiNestedMapping: {
   ...
 },
 ```
+
+## User defined types
```
Comment (sam-goldman) [inline docs/variables.md:318]: nit: User Defined Types -> [User-Defined Value Types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (including the embedded link to the Solidity docs)

File: docs/variables.md [hunk 3]
```diff
@@ -313,3 +314,17 @@ myMultiNestedMapping: {
...
+ChugSplash treats your user defined types as if they were their underlying types.
```
Comment (sam-goldman) [inline docs/variables.md:319]: > nit: User Defined Types -> User-Defined Value Types

**Comment by sam-goldman** [comment]
Looks good! Just a couple nits
